contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1236 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity 0.8.4;
+
+import "./IStaking.sol";
+import "./StakingProviders.sol";
+import "./IApplication.sol";
+import "../token/T.sol";
+import "../vending/VendingMachine.sol";
+import "../utils/PercentUtils.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+/// @notice Meta staking contract. 3 roles: app, meta staking, app manager
+contract TokenStaking is Ownable, IStaking {
+    using SafeERC20 for T;
+    using PercentUtils for uint256;
+    using SafeCast for uint256;
+
+    enum StakingProvider {
+        NU,
+        KEEP,
+        T
+    }
+
+    struct OperatorInfo {
+        address owner;
+        address payable beneficiary;
+        address authorizer;
+        mapping(address => AppAuthorization) authorizations;
+        address[] authorizedApplications;
+        uint96 nuStake;
+        uint96 keepStake;
```

Comment:
Worth dropping a comment for future-us and other contributors that these amounts are denominated in T.

Reply:
Will rename it to `keepInTStake` and `nuInTStake` or something similar